### ISSUE TITLE
Envoy infinite loops requests to itself, causing 4x memory usage

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#14242]
https://snapshot.raintank.io/dashboard/snapshot/jOGp3TQdrrCQcjzdaFd6rKXbylbplzw3?panelId=72&fullscreen&orgId=2

Leaks about 1GB  (23 -> 24 gb) over a day period, but seems to level off after that.

This is running from master.

I switched half of the proxy images over to 1.1.7 and didn't see any real changes, which means this is likely not a regression in master.

@silentdai I think you worked on something like this before, any other useful info we can get? I can give you access to the cluster too

@mandarjog

**Comment by howardjohn** [comment]
Also I replaced Pilot 1.1.7 and nothing change so I think its not a regression

**Comment by howardjohn** [comment]
Some more updates: https://snapshot.raintank.io/dashboard/snapshot/tobGjIQYgDSMnk8zajn7BpWfbCaaUsEC

See Memory increase on left, each namespace (40 pods each) memory increases by 1-5mb per 10 min. So not that much, but pretty steady

What is a lot more troublesome is the difference between namespaces. Each namespace contains the same exact setup, set service-graph06 uses 5GB total and service-graph02 doesn't even use 2gb. Looking at the resource usage per pod, (https://snapshot.raintank.io/dashboard/snapshot/imsNOlEeXWTzTODN6WWwJNLdVrELReJY?refresh=10s&panelId=10&fullscreen&orgId=2), we see this is because some pods use high memory usage while others do not. This level does not change -- it seems at start envoy decides how much memory to use and never changes it.

**Comment by howardjohn** [comment]
@mandarjog @silentdai @PiotrSikora after looking more into this I think this is worse than just a really slow leak, please see above. What information do we need to get to get to the root cause of this?

**Comment by howardjohn** [comment]
Using 275mb:
```
{
 "allocated": "15558632",
 "heap_size": "261095424",
 "pageheap_unmapped": "0",
 "pageheap_free": "185860096",
 "total_thread_cache": "27442400"
}
```

Using 32mb:
```
{
 "allocated": "12422256",
 "heap_size": "17825792",
 "pageheap_unmapped": "0",
 "pageheap_free": "696320",
 "total_thread_cache": "3082000"
}
```

Stats for 275mb one:
```
cluster.xds-grpc.assignment_stale: 0
cluster.xds-grpc.assignment_timeout_received: 0
cluster.xds-grpc.bind_errors: 0
cluster.xds-grpc.circuit_breakers.default.cx_open: 0
cluster.xds-grpc.circuit_breakers.default.cx_pool_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_pending_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_retry_open: 0
cluster.xds-grpc.circuit_breakers.high.cx_open: 0
cluster.xds-grpc.circuit_breakers.high.cx_pool_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_pending_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_retry_open: 0
cluster.xds-grpc.http2.header_overflow: 0
cluster.xds-grpc.http2.headers_cb_no_stream: 0
cluster.xds-grpc.http2.rx_messaging_error: 0
cluster.xds-grpc.http2.rx_reset: 0
cluster.xds-grpc.http2.too_many_header_frames: 0
cluster.xds-grpc.http2.trailers: 0
cluster.xds-grpc.http2.tx_reset: 0
cluster.xds-grpc.internal.upstream_rq_200: 105
cluster.xds-grpc.internal.upstream_rq_2xx: 105
cluster.xds-grpc.internal.upstream_rq_503: 15
cluster.xds-grpc.internal.upstream_rq_5xx: 15
cluster.xds-grpc.internal.upstream_rq_completed: 120
cluster.xds-grpc.lb_healthy_panic: 1
cluster.xds-grpc.lb_local_cluster_not_ok: 0
cluster.xds-grpc.lb_recalculate_zone_structures: 0
cluster.xds-grpc.lb_subsets_active: 0
cluster.xds-grpc.lb_subsets_created: 0
cluster.xds-grpc.lb_subsets_fallback: 0
cluster.xds-grpc.lb_subsets_fallback_panic: 0
cluster.xds-grpc.lb_subsets_removed: 0
cluster.xds-grpc.lb_subsets_selected: 0
cluster.xds-grpc.lb_zone_cluster_too_small: 0
cluster.xds-grpc.lb_zone_no_capacity_left: 0
cluster.xds-grpc.lb_zone_number_differs: 0
cluster.xds-grpc.lb_zone_routing_all_directly: 0
cluster.xds-grpc.lb_zone_routing_cross_zone: 0
cluster.xds-grpc.lb_zone_routing_sampled: 0
cluster.xds-grpc.max_host_weight: 1
cluster.xds-grpc.membership_change: 1
cluster.xds-grpc.membership_degraded: 0
cluster.xds-grpc.membership_excluded: 0
cluster.xds-grpc.membership_healthy: 1
cluster.xds-grpc.membership_total: 1
cluster.xds-grpc.original_dst_host_invalid: 0
cluster.xds-grpc.retry_or_shadow_abandoned: 0
cluster.xds-grpc.update_attempt: 580
cluster.xds-grpc.update_empty: 0
cluster.xds-grpc.update_failure: 0
cluster.xds-grpc.update_no_rebuild: 579
cluster.xds-grpc.update_success: 580
cluster.xds-grpc.upstream_cx_active: 1
cluster.xds-grpc.upstream_cx_close_notify: 92
cluster.xds-grpc.upstream_cx_connect_attempts_exceeded: 0
cluster.xds-grpc.upstream_cx_connect_fail: 14
cluster.xds-grpc.upstream_cx_connect_timeout: 2
cluster.xds-grpc.upstream_cx_destroy: 0
cluster.xds-grpc.upstream_cx_destroy_local: 0
cluster.xds-grpc.upstream_cx_destroy_local_with_active_rq: 0
cluster.xds-grpc.upstream_cx_destroy_remote: 0
cluster.xds-grpc.upstream_cx_destroy_remote_with_active_rq: 104
cluster.xds-grpc.upstream_cx_destroy_with_active_rq: 104
cluster.xds-grpc.upstream_cx_http1_total: 0
cluster.xds-grpc.upstream_cx_http2_total: 119
cluster.xds-grpc.upstream_cx_idle_timeout: 0
cluster.xds-grpc.upstream_cx_max_requests: 0
cluster.xds-grpc.upstream_cx_none_healthy: 1
cluster.xds-grpc.upstream_cx_overflow: 0
cluster.xds-grpc.upstream_cx_pool_overflow: 0
cluster.xds-grpc.upstream_cx_protocol_error: 0
cluster.xds-grpc.upstream_cx_rx_bytes_buffered: 13
cluster.xds-grpc.upstream_cx_rx_bytes_total: 863874305
cluster.xds-grpc.upstream_cx_total: 119
cluster.xds-grpc.upstream_cx_tx_bytes_buffered: 0
cluster.xds-grpc.upstream_cx_tx_bytes_total: 98623533
cluster.xds-grpc.upstream_flow_control_backed_up_total: 0
cluster.xds-grpc.upstream_flow_control_drained_total: 0
cluster.xds-grpc.upstream_flow_control_paused_reading_total: 0
cluster.xds-grpc.upstream_flow_control_resumed_reading_total: 0
cluster.xds-grpc.upstream_internal_redirect_failed_total: 0
cluster.xds-grpc.upstream_internal_redirect_succeeded_total: 0
cluster.xds-grpc.upstream_rq_200: 105
cluster.xds-grpc.upstream_rq_2xx: 105
cluster.xds-grpc.upstream_rq_503: 15
cluster.xds-grpc.upstream_rq_5xx: 15
cluster.xds-grpc.upstream_rq_active: 1
cluster.xds-grpc.upstream_rq_cancelled: 0
cluster.xds-grpc.upstream_rq_completed: 120
cluster.xds-grpc.upstream_rq_maintenance_mode: 0
cluster.xds-grpc.upstream_rq_pending_active: 0
cluster.xds-grpc.upstream_rq_pending_failure_eject: 118
cluster.xds-grpc.upstream_rq_pending_overflow: 0
cluster.xds-grpc.upstream_rq_pending_total: 119
cluster.xds-grpc.upstream_rq_per_try_timeout: 0
cluster.xds-grpc.upstream_rq_retry: 0
cluster.xds-grpc.upstream_rq_retry_overflow: 0
cluster.xds-grpc.upstream_rq_retry_success: 0
cluster.xds-grpc.upstream_rq_rx_reset: 0
cluster.xds-grpc.upstream_rq_timeout: 0
cluster.xds-grpc.upstream_rq_total: 105
cluster.xds-grpc.upstream_rq_tx_reset: 0
cluster.xds-grpc.version: 0
cluster_manager.active_clusters: 59
cluster_manager.cds.update_attempt: 6361
cluster_manager.cds.update_failure: 118
cluster_manager.cds.update_rejected: 0
cluster_manager.cds.update_success: 6242
cluster_manager.cds.version: 14600167950670378018
cluster_manager.cluster_added: 59
cluster_manager.cluster_modified: 1
cluster_manager.cluster_removed: 0
cluster_manager.cluster_updated: 166
cluster_manager.cluster_updated_via_merge: 0
cluster_manager.update_merge_cancelled: 0
cluster_manager.update_out_of_merge_window: 47
cluster_manager.warming_clusters: 0
http_mixer_filter.total_check_cache_hit_accepts: 0
http_mixer_filter.total_check_cache_hit_denies: 0
http_mixer_filter.total_check_cache_hits: 0
http_mixer_filter.total_check_cache_misses: 0
http_mixer_filter.total_check_calls: 0
http_mixer_filter.total_quota_cache_hit_accepts: 0
http_mixer_filter.total_quota_cache_hit_denies: 0
http_mixer_filter.total_quota_cache_hits: 0
http_mixer_filter.total_quota_cache_misses: 0
http_mixer_filter.total_quota_calls: 0
http_mixer_filter.total_remote_call_cancellations: 0
http_mixer_filter.total_remote_call_other_errors: 0
http_mixer_filter.total_remote_call_retries: 0
http_mixer_filter.total_remote_call_send_errors: 0
http_mixer_filter.total_remote_call_successes: 0
http_mixer_filter.total_remote_call_timeouts: 0
http_mixer_filter.total_remote_calls: 0
http_mixer_filter.total_remote_check_accepts: 0
http_mixer_filter.total_remote_check_calls: 0
http_mixer_filter.total_remote_check_denies: 0
http_mixer_filter.total_remote_quota_accepts: 0
http_mixer_filter.total_remote_quota_calls: 0
http_mixer_filter.total_remote_quota_denies: 0
http_mixer_filter.total_remote_quota_prefetch_calls: 0
http_mixer_filter.total_remote_report_calls: 633672
http_mixer_filter.total_remote_report_other_errors: 288
http_mixer_filter.total_remote_report_send_errors: 0
http_mixer_filter.total_remote_report_successes: 633384
http_mixer_filter.total_remote_report_timeouts: 0
http_mixer_filter.total_report_calls: 37077759
listener_manager.lds.update_attempt: 6349
listener_manager.lds.update_failure: 118
listener_manager.lds.update_rejected: 0
listener_manager.lds.update_success: 6230
listener_manager.lds.version: 14600167950670378018
listener_manager.listener_added: 31
listener_manager.listener_create_failure: 0
listener_manager.listener_create_success: 86
listener_manager.listener_modified: 12
listener_manager.listener_removed: 0
listener_manager.total_listeners_active: 31
listener_manager.total_listeners_draining: 0
listener_manager.total_listeners_warming: 0
server.concurrency: 2
server.days_until_first_cert_expiring: 87
server.debug_assertion_failures: 0
server.hot_restart_epoch: 0
server.live: 1
server.memory_allocated: 15726048
server.memory_heap_size: 261095424
server.parent_connections: 0
server.total_connections: 76
server.uptime: 173922
server.version: 5516111
server.watchdog_mega_miss: 0
server.watchdog_miss: 95
tcp_mixer_filter.total_check_cache_hit_accepts: 0
tcp_mixer_filter.total_check_cache_hit_denies: 0
tcp_mixer_filter.total_check_cache_hits: 0
tcp_mixer_filter.total_check_cache_misses: 0
tcp_mixer_filter.total_check_calls: 0
tcp_mixer_filter.total_quota_cache_hit_accepts: 0
tcp_mixer_filter.total_quota_cache_hit_denies: 0
tcp_mixer_filter.total_quota_cache_hits: 0
tcp_mixer_filter.total_quota_cache_misses: 0
tcp_mixer_filter.total_quota_calls: 0
tcp_mixer_filter.total_remote_call_cancellations: 0
tcp_mixer_filter.total_remote_call_other_errors: 0
tcp_mixer_filter.total_remote_call_retries: 0
tcp_mixer_filter.total_remote_call_send_errors: 0
tcp_mixer_filter.total_remote_call_successes: 0
tcp_mixer_filter.total_remote_call_timeouts: 0
tcp_mixer_filter.total_remote_calls: 0
tcp_mixer_filter.total_remote_check_accepts: 0
tcp_mixer_filter.total_remote_check_calls: 0
tcp_mixer_filter.total_remote_check_denies: 0
tcp_mixer_filter.total_remote_quota_accepts: 0
tcp_mixer_filter.total_remote_quota_calls: 0
tcp_mixer_filter.total_remote_quota_denies: 0
tcp_mixer_filter.total_remote_quota_prefetch_calls: 0
tcp_mixer_filter.total_remote_report_calls: 0
tcp_mixer_filter.total_remote_report_other_errors: 0
tcp_mixer_filter.total_remote_report_send_errors: 0
tcp_mixer_filter.total_remote_report_successes: 0
tcp_mixer_filter.total_remote_report_timeouts: 0
tcp_mixer_filter.total_report_calls: 0
cluster.xds-grpc.upstream_cx_connect_ms: P0(nan,0) P25(nan,0) P50(nan,0) P75(nan,0) P90(nan,1.08125) P95(nan,12.75) P99(nan,1047.5) P99.5(nan,1073.75) P99.9(nan,1094.75) P100(nan,1100)
cluster.xds-grpc.upstream_cx_length_ms: P0(nan,0) P25(nan,1.61522e+06) P50(nan,1.73846e+06) P75(nan,1.87105e+06) P90(nan,1.95083e+06) P95(nan,1.97542e+06) P99(nan,1.99508e+06) P99.5(nan,1.99754e+06) P99.9(nan,1.99951e+06) P100(nan,2e+06)
```

Stats for 32 mb one:
```
cluster.xds-grpc.assignment_stale: 0
cluster.xds-grpc.assignment_timeout_received: 0
cluster.xds-grpc.bind_errors: 0
cluster.xds-grpc.circuit_breakers.default.cx_open: 0
cluster.xds-grpc.circuit_breakers.default.cx_pool_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_pending_open: 0
cluster.xds-grpc.circuit_breakers.default.rq_retry_open: 0
cluster.xds-grpc.circuit_breakers.high.cx_open: 0
cluster.xds-grpc.circuit_breakers.high.cx_pool_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_pending_open: 0
cluster.xds-grpc.circuit_breakers.high.rq_retry_open: 0
cluster.xds-grpc.http2.header_overflow: 0
cluster.xds-grpc.http2.headers_cb_no_stream: 0
cluster.xds-grpc.http2.rx_messaging_error: 0
cluster.xds-grpc.http2.rx_reset: 0
cluster.xds-grpc.http2.too_many_header_frames: 0
cluster.xds-grpc.http2.trailers: 0
cluster.xds-grpc.http2.tx_reset: 0
cluster.xds-grpc.internal.upstream_rq_200: 105
cluster.xds-grpc.internal.upstream_rq_2xx: 105
cluster.xds-grpc.internal.upstream_rq_503: 14
cluster.xds-grpc.internal.upstream_rq_5xx: 14
cluster.xds-grpc.internal.upstream_rq_completed: 119
cluster.xds-grpc.lb_healthy_panic: 1
cluster.xds-grpc.lb_local_cluster_not_ok: 0
cluster.xds-grpc.lb_recalculate_zone_structures: 0
cluster.xds-grpc.lb_subsets_active: 0
cluster.xds-grpc.lb_subsets_created: 0
cluster.xds-grpc.lb_subsets_fallback: 0
cluster.xds-grpc.lb_subsets_fallback_panic: 0
cluster.xds-grpc.lb_subsets_removed: 0
cluster.xds-grpc.lb_subsets_selected: 0
cluster.xds-grpc.lb_zone_cluster_too_small: 0
cluster.xds-grpc.lb_zone_no_capacity_left: 0
cluster.xds-grpc.lb_zone_number_differs: 0
cluster.xds-grpc.lb_zone_routing_all_directly: 0
cluster.xds-grpc.lb_zone_routing_cross_zone: 0
cluster.xds-grpc.lb_zone_routing_sampled: 0
cluster.xds-grpc.max_host_weight: 1
cluster.xds-grpc.membership_change: 1
cluster.xds-grpc.membership_degraded: 0
cluster.xds-grpc.membership_excluded: 0
cluster.xds-grpc.membership_healthy: 1
cluster.xds-grpc.membership_total: 1
cluster.xds-grpc.original_dst_host_invalid: 0
cluster.xds-grpc.retry_or_shadow_abandoned: 0
cluster.xds-grpc.update_attempt: 580
cluster.xds-grpc.update_empty: 0
cluster.xds-grpc.update_failure: 0
cluster.xds-grpc.update_no_rebuild: 579
cluster.xds-grpc.update_success: 580
cluster.xds-grpc.upstream_cx_active: 1
cluster.xds-grpc.upstream_cx_close_notify: 92
cluster.xds-grpc.upstream_cx_connect_attempts_exceeded: 0
cluster.xds-grpc.upstream_cx_connect_fail: 13
cluster.xds-grpc.upstream_cx_connect_timeout: 3
cluster.xds-grpc.upstream_cx_destroy: 0
cluster.xds-grpc.upstream_cx_destroy_local: 0
cluster.xds-grpc.upstream_cx_destroy_local_with_active_rq: 0
cluster.xds-grpc.upstream_cx_destroy_remote: 0
cluster.xds-grpc.upstream_cx_destroy_remote_with_active_rq: 104
cluster.xds-grpc.upstream_cx_destroy_with_active_rq: 104
cluster.xds-grpc.upstream_cx_http1_total: 0
cluster.xds-grpc.upstream_cx_http2_total: 118
cluster.xds-grpc.upstream_cx_idle_timeout: 0
cluster.xds-grpc.upstream_cx_max_requests: 0
cluster.xds-grpc.upstream_cx_none_healthy: 1
cluster.xds-grpc.upstream_cx_overflow: 0
cluster.xds-grpc.upstream_cx_pool_overflow: 0
cluster.xds-grpc.upstream_cx_protocol_error: 0
cluster.xds-grpc.upstream_cx_rx_bytes_buffered: 30
cluster.xds-grpc.upstream_cx_rx_bytes_total: 842249742
cluster.xds-grpc.upstream_cx_total: 118
cluster.xds-grpc.upstream_cx_tx_bytes_buffered: 0
cluster.xds-grpc.upstream_cx_tx_bytes_total: 98876520
cluster.xds-grpc.upstream_flow_control_backed_up_total: 0
cluster.xds-grpc.upstream_flow_control_drained_total: 0
cluster.xds-grpc.upstream_flow_control_paused_reading_total: 0
cluster.xds-grpc.upstream_flow_control_resumed_reading_total: 0
cluster.xds-grpc.upstream_internal_redirect_failed_total: 0
cluster.xds-grpc.upstream_internal_redirect_succeeded_total: 0
cluster.xds-grpc.upstream_rq_200: 105
cluster.xds-grpc.upstream_rq_2xx: 105
cluster.xds-grpc.upstream_rq_503: 14
cluster.xds-grpc.upstream_rq_5xx: 14
cluster.xds-grpc.upstream_rq_active: 1
cluster.xds-grpc.upstream_rq_cancelled: 0
cluster.xds-grpc.upstream_rq_completed: 119
cluster.xds-grpc.upstream_rq_maintenance_mode: 0
cluster.xds-grpc.upstream_rq_pending_active: 0
cluster.xds-grpc.upstream_rq_pending_failure_eject: 117
cluster.xds-grpc.upstream_rq_pending_overflow: 0
cluster.xds-grpc.upstream_rq_pending_total: 118
cluster.xds-grpc.upstream_rq_per_try_timeout: 0
cluster.xds-grpc.upstream_rq_retry: 0
cluster.xds-grpc.upstream_rq_retry_overflow: 0
cluster.xds-grpc.upstream_rq_retry_success: 0
cluster.xds-grpc.upstream_rq_rx_reset: 0
cluster.xds-grpc.upstream_rq_timeout: 0
cluster.xds-grpc.upstream_rq_total: 105
cluster.xds-grpc.upstream_rq_tx_reset: 0
cluster.xds-grpc.version: 0
cluster_manager.active_clusters: 58
cluster_manager.cds.update_attempt: 6352
cluster_manager.cds.update_failure: 117
cluster_manager.cds.update_rejected: 0
cluster_manager.cds.update_success: 6234
cluster_manager.cds.version: 14600167950670378018
cluster_manager.cluster_added: 58
cluster_manager.cluster_modified: 1
cluster_manager.cluster_removed: 0
cluster_manager.cluster_updated: 146
cluster_manager.cluster_updated_via_merge: 0
cluster_manager.update_merge_cancelled: 0
cluster_manager.update_out_of_merge_window: 47
cluster_manager.warming_clusters: 0
http_mixer_filter.total_check_cache_hit_accepts: 0
http_mixer_filter.total_check_cache_hit_denies: 0
http_mixer_filter.total_check_cache_hits: 0
http_mixer_filter.total_check_cache_misses: 0
http_mixer_filter.total_check_calls: 0
http_mixer_filter.total_quota_cache_hit_accepts: 0
http_mixer_filter.total_quota_cache_hit_denies: 0
http_mixer_filter.total_quota_cache_hits: 0
http_mixer_filter.total_quota_cache_misses: 0
http_mixer_filter.total_quota_calls: 0
http_mixer_filter.total_remote_call_cancellations: 0
http_mixer_filter.total_remote_call_other_errors: 0
http_mixer_filter.total_remote_call_retries: 0
http_mixer_filter.total_remote_call_send_errors: 0
http_mixer_filter.total_remote_call_successes: 0
http_mixer_filter.total_remote_call_timeouts: 0
http_mixer_filter.total_remote_calls: 0
http_mixer_filter.total_remote_check_accepts: 0
http_mixer_filter.total_remote_check_calls: 0
http_mixer_filter.total_remote_check_denies: 0
http_mixer_filter.total_remote_quota_accepts: 0
http_mixer_filter.total_remote_quota_calls: 0
http_mixer_filter.total_remote_quota_denies: 0
http_mixer_filter.total_remote_quota_prefetch_calls: 0
http_mixer_filter.total_remote_report_calls: 0
http_mixer_filter.total_remote_report_other_errors: 0
http_mixer_filter.total_remote_report_send_errors: 0
http_mixer_filter.total_remote_report_successes: 0
http_mixer_filter.total_remote_report_timeouts: 0
http_mixer_filter.total_report_calls: 0
listener_manager.lds.update_attempt: 6344
listener_manager.lds.update_failure: 117
listener_manager.lds.update_rejected: 0
listener_manager.lds.update_success: 6226
listener_manager.lds.version: 14600167950670378018
listener_manager.listener_added: 31
listener_manager.listener_create_failure: 0
listener_manager.listener_create_success: 82
listener_manager.listener_modified: 10
listener_manager.listener_removed: 0
listener_manager.total_listeners_active: 31
listener_manager.total_listeners_draining: 0
listener_manager.total_listeners_warming: 0
server.concurrency: 2
server.days_until_first_cert_expiring: 87
server.debug_assertion_failures: 0
server.hot_restart_epoch: 0
server.live: 1
server.memory_allocated: 12388304
server.memory_heap_size: 17825792
server.parent_connections: 0
server.total_connections: 7
server.uptime: 173821
server.version: 5516111
server.watchdog_mega_miss: 0
server.watchdog_miss: 101
tcp_mixer_filter.total_check_cache_hit_accepts: 0
tcp_mixer_filter.total_check_cache_hit_denies: 0
tcp_mixer_filter.total_check_cache_hits: 0
tcp_mixer_filter.total_check_cache_misses: 0
tcp_mixer_filter.total_check_calls: 0
tcp_mixer_filter.total_quota_cache_hit_accepts: 0
tcp_mixer_filter.total_quota_cache_hit_denies: 0
tcp_mixer_filter.total_quota_cache_hits: 0
tcp_mixer_filter.total_quota_cache_misses: 0
tcp_mixer_filter.total_quota_calls: 0
tcp_mixer_filter.total_remote_call_cancellations: 0
tcp_mixer_filter.total_remote_call_other_errors: 0
tcp_mixer_filter.total_remote_call_retries: 0
tcp_mixer_filter.total_remote_call_send_errors: 0
tcp_mixer_filter.total_remote_call_successes: 0
tcp_mixer_filter.total_remote_call_timeouts: 0
tcp_mixer_filter.total_remote_calls: 0
tcp_mixer_filter.total_remote_check_accepts: 0
tcp_mixer_filter.total_remote_check_calls: 0
tcp_mixer_filter.total_remote_check_denies: 0
tcp_mixer_filter.total_remote_quota_accepts: 0
tcp_mixer_filter.total_remote_quota_calls: 0
tcp_mixer_filter.total_remote_quota_denies: 0
tcp_mixer_filter.total_remote_quota_prefetch_calls: 0
tcp_mixer_filter.total_remote_report_calls: 69131
tcp_mixer_filter.total_remote_report_other_errors: 15
tcp_mixer_filter.total_remote_report_send_errors: 0
tcp_mixer_filter.total_remote_report_successes: 69116
tcp_mixer_filter.total_remote_report_timeouts: 0
tcp_mixer_filter.total_report_calls: 86492
cluster.xds-grpc.upstream_cx_connect_ms: P0(nan,0) P25(nan,0) P50(nan,0) P75(nan,0) P90(nan,1.07) P95(nan,10.875) P99(nan,21.95) P99.5(nan,3147.5) P99.9(nan,3189.5) P100(nan,3200)
cluster.xds-grpc.upstream_cx_length_ms: P0(nan,0) P25(nan,1.61548e+06) P50(nan,1.746e+06) P75(nan,1.86058e+06) P90(nan,1.93842e+06) P95(nan,1.96921e+06) P99(nan,1.99384e+06) P99.5(nan,1.99692e+06) P99.9(nan,1.99938e+06) P100(nan,2e+06)
```


Seems the small one only is reporting tcp stats (tcp_mixer_filter.total_report_calls: 86492) and the other http stats (http_mixer_filter.total_report_calls: 37077759). The other stats are mostly identical. Will continue looking to see if this is true for all cases

**Comment by howardjohn** [comment]
![2019-24-05_10-39-30](https://user-images.githubusercontent.com/623453/58346600-41cc5f00-7e10-11e9-8050-b6e68f31d1d0.png)
First I went into each pod and did `pkill -9 envoy`. We see memory usage drops back to normal levels. Then I `kubectl delete pods --all`, and when they come up, they use sporadic memory amounts again. So killing just envoy fixes it, killing pod does not

**Comment by silentdai** [comment]
```
Using 275mb:{ "allocated": "15558632", "heap_size": "261095424",
```
Heap is 261Mb vs allocated is 15Mb. Indicating peak usage 261Mb.
It's unknown why this happens. Will investigate.

However, os should be able to reclaim allocated in recent envoy.

**Comment by silentdai** [comment]
`kill -9 envoy` works because envoy is not respawned. 
If the pod recreation bring up the huge usage again, it's probably not the leak due to config fluctuation.

**Comment by howardjohn** [comment]
@silentdai I don't think it is a leak due to config, the memory starts high and stays high. The leak I think is a different issue, and its like a couple MB an hour so not a big deal.

And when we `kill -9 envoy` pilot-agent recreates envoy

**Comment by howardjohn** [comment]
Slightly better graph: https://snapshot.raintank.io/dashboard/snapshot/MznqDhmiKlGAOEXDC2GDzASvvhCyIGrT?refresh=10s&panelId=4&fullscreen&orgId=2

At 10:23 I started killing envoys, at 10:37 I started killing pods

**Comment by silentdai** [comment]
As the `kill -9` spawns envoy, I would suspect a heavy allocation at the starting of envoy. 
Do we have `listener_manager.lds.update_attempt` in normal envoy and the leaking envoy?
Any other xDS worth noticing but lds is good start.

**Comment by howardjohn** [comment]
@silentdai from the two stats I posted above, update_attempts are both almost the same at ~6000. One was using 35mb one was using 275mb

**Comment by silentdai** [comment]
Create the overload action issue to describe how to reclaim heap memory.
#14366

**Comment by howardjohn** [comment]
It seems like this is due to management ports. Using service graph we have port 8080 taken, so normally get a lot of `Omitting listener for management address 10.28.0.125_8080 due to collision with service listener 10.28.0.125_8080` in the logs. There seems to be some correlation to these listeners getting omitted and the memory usage, but I haven't pinpointed it yet (it doesn't happen every time which makes it harder). I think it is some sort of race condition in pilot, as it only happens when a lot of configs are changed at once. It seems most of the time `istioctl ps` will show missing listeners for  the management pods for the ones with high memory.

High memory ones also tend to show
```
* failed checking application ports. listeners="0.0.0.0:15090","10.0.44.82:15030","10.0.44.82:15443","10.0.41.147:8080","10.0.44.82:15020","10.0.46.146:443","10.0.33.118:8080","10.0.32.1:443","10.0.38.240:8080","0.0.0.0:9090","10.0.39.243:8080","10.0.44.19:443","10.0.32.2:443","10.0.44.82:443","10.0.39.74:15011","10.0.44.92:443","10.0.44.82:15031","10.0.46.28:42422","0.0.0.0:10250","10.0.44.10:80","10.0.44.82:15032","10.0.44.177:5473","10.0.44.82:15029","10.0.32.48:9090","10.0.45.203:9100","10.0.38.252:8080","10.0.32.10:53","10.0.44.82:31400","0.0.0.0:3000","0.0.0.0:8080","0.0.0.0:8060","0.0.0.0:15004","0.0.0.0:15014","0.0.0.0:9091","0.0.0.0:80","0.0.0.0:9901","0.0.0.0:15010","0.0.0.0:15001"
```
at startup.

Still investigating this, mostly putting here as notes to self

**Comment by howardjohn** [comment]
```
--- Pilot Clusters
+++ Envoy Clusters
@@ -13,80 +13,14 @@
             "type": "ORIGINAL_DST",
             "connectTimeout": "10s",
             "lbPolicy": "ORIGINAL_DST_LB"
          }
       },
       {
          "cluster": {
-            "name": "inbound|15020|mgmt-15020|mgmtCluster",
-            "type": "STATIC",
-            "connectTimeout": "10s",
-            "loadAssignment": {
-               "clusterName": "inbound|15020|mgmt-15020|mgmtCluster",
-               "endpoints": [
-                  {
-                     "lbEndpoints": [
-                        {
-                           "endpoint": {
-                              "address": {
-                                 "socketAddress": {
-                                    "address": "127.0.0.1",
-                                    "portValue": 15020
-                                 }
-                              }
-                           }
-                        }
-                     ]
-                  }
-               ]
-            },
-            "circuitBreakers": {
-               "thresholds": [
-                  {
-
-                  }
-               ]
-            }
-         }
-      },
-      {
-         "cluster": {
-            "name": "inbound|8080|mgmt-8080|mgmtCluster",
-            "type": "STATIC",
-            "connectTimeout": "10s",
-            "loadAssignment": {
-               "clusterName": "inbound|8080|mgmt-8080|mgmtCluster",
-               "endpoints": [
-                  {
-                     "lbEndpoints": [
-                        {
-                           "endpoint": {
-                              "address": {
-                                 "socketAddress": {
-                                    "address": "127.0.0.1",
-                                    "portValue": 8080
-                                 }
-                              }
-                           }
-                        }
-                     ]
-                  }
-               ]
-            },
-            "circuitBreakers": {
-               "thresholds": [
-                  {
-
-                  }
-               ]
-            }
-         }
-      },
-      {
-         "cluster": {
             "name": "outbound|10250||istio-kubelet.kube-system.svc.cluster.local",
             "type": "ORIGINAL_DST",
             "connectTimeout": "10s",
             "lbPolicy": "ORIGINAL_DST_LB",
             "circuitBreakers": {
                "thresholds": [
                   {

--- Pilot Listeners
+++ Envoy Listeners
@@ -2945,68 +2945,14 @@
             "deprecatedV1": {
                "bindToPort": false
             }
          }
       },
       {
          "listener": {
-            "name": "10.28.1.158_15020",
-            "address": {
-               "socketAddress": {
-                  "address": "10.28.1.158",
-                  "portValue": 15020
-               }
-            },
-            "filterChains": [
-               {
-                  "filters": [
-                     {
-                        "name": "envoy.tcp_proxy",
-                        "config": {
-                              "cluster": "inbound|15020|mgmt-15020|mgmtCluster",
-                              "stat_prefix": "inbound|15020|mgmt-15020|mgmtCluster"
-                           }
-                     }
-                  ]
-               }
-            ],
-            "deprecatedV1": {
-               "bindToPort": false
-            }
-         }
-      },
-      {
-         "listener": {
-            "name": "10.28.1.158_8080",
-            "address": {
-               "socketAddress": {
-                  "address": "10.28.1.158",
-                  "portValue": 8080
-               }
-            },
-            "filterChains": [
-               {
-                  "filters": [
-                     {
-                        "name": "envoy.tcp_proxy",
-                        "config": {
-                              "cluster": "inbound|8080|mgmt-8080|mgmtCluster",
-                              "stat_prefix": "inbound|8080|mgmt-8080|mgmtCluster"
-                           }
-                     }
-                  ]
-               }
-            ],
-            "deprecatedV1": {
-               "bindToPort": false
-            }
-         }
-      },
-      {
-         "listener": {
             "name": "virtual",
             "address": {
                "socketAddress": {
                   "address": "0.0.0.0",
                   "portValue": 15001
                }
             },
```

proxy-status diff of high memory usage

**Comment by howardjohn** [comment]
Ran git bisect, producing this cool graph: https://snapshot.raintank.io/dashboard/snapshot/wolg2L0yxvs77Sfvig44rR6HT6r8CDeT?panelId=10&fullscreen&orgId=2

Culprit is apparently fc55b31069986ab6d0f6acecfaf7b3504dcd294a, which is pretty odd, since this change is already on release-1.1. Maybe need to explicitly turn the flag on and retry this..

**Comment by howardjohn** [comment]
With flag on:
1.1.3 is also broken. This is when the feature was first introduced. In 1.1.4 it was extended to listeners (HTTPS), so it being broken means the virtual host stuff is the culprit.

With REGISTRY_ONLY set, this issue is NOT present, which narrows it down further to the added wildcard route.

**Comment by howardjohn** [comment]
Another nice graph of the slow leak https://snapshot.raintank.io/dashboard/snapshot/rol0aNcIs2I45RpfHrW2gdDOg9KMYW1z?panelId=72&fullscreen&orgId=2, over a 1 week period

**Comment by howardjohn** [comment]
I tried really hard to isolate this to envoy, and created the same config that was causing this issue (using https://github.com/howardjohn/file-based-istio), and they all used the same small amount of memory. It is possible this just means the issue is in Envoy's code for a real XDS connection, or something to do with the sequence of pushes.

I am not sure what the next steps to debug this are @silentdai @PiotrSikora 

A quick summary: creating a route that routes to PassthroughCluster (original dest LB) causes Envoy to use 2-5x memory in nondeterministically and immediately (not a leak). This seems to only happen when creating many pods at once, suggesting a sequencing issue with Pilot. If we `pkill -9 envoy` the memory always drops back to the right level, but removing the pod does not.

**Comment by howardjohn** [comment]
[envoy-heap.tar.gz](https://github.com/istio/istio/files/3288575/envoy-heap.tar.gz)


Assuming in path /tmp/envoy, run with `PPROF_BINARY_PATH=/tmp/envoy/lib pprof -http :8080 /tmp/envoy/*` (if there is a better way please tell me, I just copied a bunch of stuff from the pod and threw it in the lib folder)

**Comment by howardjohn** [comment]
Also some more info on heap vs alloc:
```
badhealth-df448f8d5-w4lrl:      11mb alloc      15mb heap       75% of heap
svc00-0-0-0-6d9d4c4f87-2bcg4:   12mb alloc      17mb heap       72% of heap
svc00-0-0-0-6d9d4c4f87-qts6q:   12mb alloc      17mb heap       72% of heap
svc00-0-0-79c56b96c5-c79ql:     55mb alloc      464mb heap      11% of heap
svc00-0-0-79c56b96c5-tnh2w:     58mb alloc      552mb heap      10% of heap
svc00-0-1-0-5d699b648c-dm666:   49mb alloc      268mb heap      18% of heap
svc00-0-1-0-5d699b648c-dmxks:   39mb alloc      270mb heap      14% of heap
svc00-0-1-b7dfd7dc-q7dfq:       51mb alloc      325mb heap      15% of heap
svc00-0-1-b7dfd7dc-xg78f:       50mb alloc      307mb heap      16% of heap
svc00-0-2-0-6bb9ff96bc-kdtb5:   53mb alloc      318mb heap      16% of heap
svc00-0-2-0-6bb9ff96bc-wmmdt:   55mb alloc      457mb heap      12% of heap
svc00-0-2-66fff7bcbc-27cw9:     57mb alloc      463mb heap      12% of heap
svc00-0-2-66fff7bcbc-9tf9k:     56mb alloc      525mb heap      10% of heap
svc00-0-3-0-65d5f65999-dvwl9:   41mb alloc      264mb heap      15% of heap
svc00-0-3-0-65d5f65999-vb6k7:   55mb alloc      466mb heap      11% of heap
svc00-0-3-687db98dd7-6xw7b:     56mb alloc      495mb heap      11% of heap
svc00-0-3-687db98dd7-kw4l6:     51mb alloc      349mb heap      14% of heap
svc00-0-4-0-bd8689bcf-kntgt:    56mb alloc      430mb heap      13% of heap
svc00-0-4-0-bd8689bcf-xvxs2:    58mb alloc      402mb heap      14% of heap
svc00-0-4-6497759f5f-45r96:     57mb alloc      452mb heap      12% of heap
svc00-0-4-6497759f5f-dmnfq:     14mb alloc      18mb heap       74% of heap
svc00-0-5-0-55b44df6d5-bkp7m:   12mb alloc      15mb heap       78% of heap
svc00-0-5-0-55b44df6d5-p2jkd:   12mb alloc      15mb heap       78% of heap
svc00-0-5-559fd46dd7-bfxh2:     49mb alloc      262mb heap      18% of heap
```

**Comment by howardjohn** [comment]
[logs.zip](https://github.com/istio/istio/files/3291491/logs.zip)
Logs of 2 good proxies and one bad proxy. Seems that it is causing an infinite loop somewhere with redirects/iptables.. Probably related to https://github.com/istio/istio/issues/14443

**Comment by howardjohn** [comment]
Some more info - the applications have a readiness probe that hits /metrics. For some reason the good proxies don't even get requests to /metrics? Or they are not reported in access log at least. Will investigate that part further but I don't think it will lead us to the root cause.

However, the bad log shows this infinite looping only until "Envoy proxy is ready" is logged by pilot-agent

So there is something with the initial config that is broken, then once all config is pushed it seems ok

**Comment by howardjohn** [comment]
[good.log](https://github.com/istio/istio/files/3291575/good.log) ahh. This is one with low memory usage but has the same infinite logging stuff.

**Comment by howardjohn** [comment]
Some don't have readiness probes. For future reference, it is these ones (1 == no readiness probe)

```
11mb    2       badhealth-df448f8d5-vbtrj
mb      0       config-change-cf8bd5659-ttzwb
mb      0       noproxy-d5574649b-cqph9
11mb    1       shell-6f85b78468-rqjbl
43mb    2       svc00-0-0-0-6d9d4c4f87-vd92w
53mb    2       svc00-0-0-0-6d9d4c4f87-xphn2
11mb    2       svc00-0-0-79c56b96c5-bpldx
11mb    2       svc00-0-0-79c56b96c5-zq248
11mb    2       svc00-0-1-0-5d699b648c-27nh6
30mb    2       svc00-0-1-0-5d699b648c-zkpcj
18mb    2       svc00-0-1-b7dfd7dc-rg9cj
11mb    2       svc00-0-1-b7dfd7dc-vn5w7
33mb    2       svc00-0-2-0-6bb9ff96bc-nxr5h
11mb    2       svc00-0-2-0-6bb9ff96bc-tbb89
11mb    2       svc00-0-2-66fff7bcbc-qgj5r
47mb    2       svc00-0-2-66fff7bcbc-wrzqd
11mb    2       svc00-0-3-0-65d5f65999-lsbdp
17mb    2       svc00-0-3-0-65d5f65999-n7qgt
38mb    2       svc00-0-3-687db98dd7-vx2v8
55mb    2       svc00-0-3-687db98dd7-w8nsq
47mb    2       svc00-0-4-0-bd8689bcf-mk69j
11mb    2       svc00-0-4-0-bd8689bcf-nch8f
11mb    2       svc00-0-4-6497759f5f-29hjt
51mb    2       svc00-0-4-6497759f5f-mgndv
11mb    1       svc00-0-5-0-55b44df6d5-frps9
11mb    1       svc00-0-5-0-55b44df6d5-w8w46
11mb    2       svc00-0-5-559fd46dd7-6mnv5
11mb    2       svc00-0-5-559fd46dd7-lgm5d
11mb    1       svc00-0-6-0-6b9dff5745-c7rdv
11mb    1       svc00-0-6-0-6b9dff5745-g56q8
11mb    2       svc00-0-6-68c68b4d4c-8wmpd
11mb    2       svc00-0-6-68c68b4d4c-zk475
11mb    1       svc00-0-7-0-6686749f85-hb8kv
11mb    1       svc00-0-7-0-6686749f85-xn5sn
11mb    2       svc00-0-7-566b788f8c-mxw49
11mb    2       svc00-0-7-566b788f8c-rk8m9
11mb    1       svc00-0-8-0-749fcf86df-4vrmb
11mb    1       svc00-0-8-0-749fcf86df-rknnc
32mb    2       svc00-0-8-66767dd787-8l4m2
48mb    2       svc00-0-8-66767dd787-mtz6z
34mb    2       svc00-0v1-64898b5fff-jx9qb
47mb    2       svc00-0v1-64898b5fff-m5z24
11mb    2       svc00-0v2-7cd94dfd46-h4pc5
45mb    2       svc00-0v2-7cd94dfd46-mw58l
```

**Comment by howardjohn** [comment]
I think @silentdai finally found the root cause

Say we have a pod with ip 10.44.4.175.

Eventually I will have listeners `10.44.4.175:8080` which is the inbound and `0.0.0.0:8080` which is outbound.

But sometimes the outbound sends first, so the request goes to `0.0.0.0` instead

It doesn't find a matching route for `10.44.4.175:8080`, so it gets directed to passthrough cluster (NEW BEHAVIOR -- it used to get 404!).

So then it goes back to the same IP, and the process is repeated until the max_retries is hit

**Comment by howardjohn** [comment]
In summary:
* There is a race condition in which inbound requests will get matched to the outbound listeners.
* These requests will then eventually be directed to the PassthroughCluster, which sends the request back to itself.
* Request will loop until it hits max retries (1024).
* This problem goes away once the inbound listeners are sent, BUT, during the startup phase the envoy ends up using a ton of memory (500mb+) and this is never fully reclaimed.

Possible approaches:
* Before routing to passthrough, check the X-Forwarded-For header. If it ends with the pod's IP, then send to blackhole instead/502. This won't fix infinite loops from sending to other ports like seen in https://github.com/istio/istio/issues/14443, and is a bit of a hack. Only works for http.
* Add support for a wildcard port for Envoy listeners. The match order will be SpecificIP:SpecificPort (inbound case), SpecificIp:WildcardPort (new listener we add), then WildcardIp. This way even before we have the inbound listener, we can match all inbound traffic and direct to blackhole cluster.

@hzxuzhonghu @rshriram ideas welcome :slightly_smiling_face:

**Comment by mandarjog** [comment]
This should be fixed by enabling filter chain separation.
Traffic coming from outside should not be subjected to outbound config.

**Comment by howardjohn** [comment]
In its current state filter chain separation won't do this. Even though the virtual listener (the one that actually binds to the port) is split, they still use the same non-binding listeners.


I added this to Envoy, which gives new meaning to port `0` (which as far as I know does nothing with bind_to_port=false), with `0` meaning "any port":
```diff
diff --git a/source/server/connection_handler_impl.cc b/source/server/connection_handler_impl.cc
index 855cdaadb..6b77e1f2d 100644
--- a/source/server/connection_handler_impl.cc
+++ b/source/server/connection_handler_impl.cc
@@ -123,6 +123,28 @@ ConnectionHandlerImpl::findActiveListenerByAddress(const Network::Address::Insta
 
+  // Next we look for matching IP but port 0
+  listener_it = std::find_if(
+      listeners_.begin(), listeners_.end(),
+      [&address](const std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerPtr>& p) {
+        return p.second->listener_ != nullptr && p.first->type() == Network::Address::Type::Ip &&
+               p.first->ip()->port() == 0 &&
+               (p.first->ip()->addressAsString().compare(address.ip()->addressAsString()) == 0);
+      });
+
+  // If there is matching IP with port 0 return it
+  if (listener_it != listeners_.end()) {
+     return listener_it->second.get();
+  }
```

Then added this to pilot
```diff
diff --git a/pilot/pkg/networking/core/v1alpha3/listener.go b/pilot/pkg/networking/core/v1alpha3/listener.go
index 1acbb9593..6af77bb2e 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener.go
@@ -213,8 +213,40 @@ func (configgen *ConfigGeneratorImpl) buildSidecarListeners(env *model.Environme
 		// TODO(silentdai): remove `if` branch once split behavior is well verified
 		if !node.IsInboundCaptureAllPorts() {
 			inbound := configgen.buildSidecarInboundListeners(env, node, push, proxyInstances)
+
+			inboundWildcards := []*xdsapi.Listener{}
+			for _, ip := range node.IPAddresses {
+				wildcardMatch := &listener.FilterChainMatch{}
+				tcpFilter := listener.Filter{
+					Name: xdsutil.TCPProxy,
+				}
+				config := &tcp_proxy.TcpProxy{
+					StatPrefix:       util.BlackHoleCluster,
+					ClusterSpecifier: &tcp_proxy.TcpProxy_Cluster{Cluster: util.BlackHoleCluster},
+				}
+				if util.IsXDSMarshalingToAnyEnabled(node) {
+					tcpFilter.ConfigType = &listener.Filter_TypedConfig{TypedConfig: util.MessageToAny(config)}
+				} else {
+					tcpFilter.ConfigType = &listener.Filter_Config{Config: util.MessageToStruct(config)}
+				}
+				inboundWildcards = append(inboundWildcards, &xdsapi.Listener{
+					Name:    ip + "_*",
+					Address: util.BuildAddress(ip, 0),
+					DeprecatedV1: &xdsapi.Listener_DeprecatedV1{
+						BindToPort: proto.BoolFalse,
+					},
+					FilterChains: []listener.FilterChain{
+						{
+							FilterChainMatch: wildcardMatch,
+							Filters:          []listener.Filter{tcpFilter},
+						},
+					},
+				})
+			}
+
 			outbound := configgen.buildSidecarOutboundListeners(env, node, push, proxyInstances)
 
+			listeners = append(listeners, inboundWildcards...)
 			listeners = append(listeners, inbound...)
 			listeners = append(listeners, outbound...)
 ```


and the problem is resolved. This could also be added to bootstrap if we wanted.

Yuchen also has a solution that is probably better (was my first time touching envoy) but I haven't tested yet: https://github.com/silentdai/envoy/pull/3/files

**Comment by hzxuzhonghu** [comment]
>Eventually I will have listeners 10.44.4.175:8080 which is the inbound and 0.0.0.0:8080 which is outbound.
But sometimes the outbound sends first, so the request goes to 0.0.0.0 instead

How could outbound listeners send first? In pilot, it first gen inbound listeners and then outbound ones.

If inbound listeners not received, the sidecar readiness probe will fail, and then k8s will not mark this pod an ready endpoint of the service, so it should not receive any requests, am i missing something?

**Comment by howardjohn** [comment]
Yes I forgot part of it, the application should have readiness probe. Then kubernetes will send requests before it is marked as ready (before inbound is sent)

**Comment by hzxuzhonghu** [comment]
oh, for the 2 doubt, i find pilot does not ignore not ready ones

```
func (c *Controller) getProxyServiceInstancesByEndpoint(endpoints v1.Endpoints, proxy *model.Proxy) []*model.ServiceInstance {
	out := make([]*model.ServiceInstance, 0)

	hostname := serviceHostname(endpoints.Name, endpoints.Namespace, c.domainSuffix)
	c.RLock()
	svc := c.servicesMap[hostname]
	c.RUnlock()

	if svc != nil {
		for _, ss := range endpoints.Subsets {
			for _, port := range ss.Ports {
				svcPort, exists := svc.Ports.Get(port.Name)
				if !exists {
					continue
				}

				// There is only one IP for kube registry
				proxyIP := proxy.IPAddresses[0]

				if hasProxyIP(ss.Addresses, proxyIP) {
					out = append(out, c.getEndpoints(proxyIP, port.Port, svcPort, svc))
				}

				if hasProxyIP(ss.NotReadyAddresses, proxyIP) {
					nrEP := c.getEndpoints(proxyIP, port.Port, svcPort, svc)
					out = append(out, nrEP)
					if c.Env != nil {
						c.Env.PushContext.Add(model.ProxyStatusEndpointNotReady, proxy.ID, proxy, "")
					}
				}
			}
		}
	}

	return out
}
```

**Comment by hzxuzhonghu** [comment]
Yes, according to func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {

we did ignore unready endpoints addresses

**Comment by howardjohn** [comment]
Does this mean that we only send the inbound listener after the it is
ready?

**Comment by hzxuzhonghu** [comment]
no, it means downstream only send request to it once it is ready (after receiving the inbound listener).

**Comment by howardjohn** [comment]
I see thanks. But there are still ways requests get sent to unready endpoints , specifically readiness probes and maybe others

**Comment by ckcd** [comment]
in istio 1.2.0, this memory increase is more serious, with test case in https://github.com/istio/istio/issues/14443 , memory of envoy will increase immediately and be OOM Killed.

**Comment by silentdai** [comment]
Forget the implementation of listener or filter chain: 
- iptables output chain is supposed to send the packet back to catpure port at the clientenvoy - serviceenvoy connection.
- the inbound listener is supposed to redirect the unknown traffic to original dst.  Notes the "unknown" could refer to various of cases.

My point is that **inbound** listener should capture the traffic to envoy, but the envoy should *NEVER* send to original dst. **Inbound listener** should only redirect to the loopback address.

WDTY?

**Comment by rshriram** [comment]
That was my suggestion as well. You simply need a orig dst port with loopback address.

**Comment by silentdai** [comment]
Might work with containerPort but not align with the efforts that deprecating containerPort. Better find a solution with containerPort and without
